File: my_solutions/container_with_most_water.js
//  O(1) space and O(N) time
function maxArea(height) {
  // Use two pointers to traverse the array
  let pointer1 = 0;
  let pointer2 = height.length - 1;
  // use a variable to keep track of the maxHeight
  let maxHeight = 0;

  while (pointer1 != pointer2) {
    //  difference is the index at pointer2 (upper bound) minus index at pointer1
    // (lower bound)
    let difference = pointer2 - pointer1;
    // find smallest value
    let smallestValue = Math.min(height[pointer2], height[pointer1]);
    // area will be calculated
    let area = difference * smallestValue;

    //  maxHeight needs to be tracked
    if (area > maxHeight) maxHeight = area;

    if (height[pointer1] < height[pointer2]) {
      pointer1++;
    } else {
      pointer2--;
    }
  }
  return maxHeight;
}

File: container_with_most_water.dart
import 'dart:math';

class Solution {
  int maxArea(List<int> height, [void Function()? name]) {
    int highest = 0;
    int p1 = 0;
    int p2 = height.length - 1;
    while (p1 != p2) {
      int difference = p2 - p1;
      int smallestNumber = min(height[p2], height[p1]);
      print(smallestNumber);
      int area = difference * smallestNumber;
      if (area > highest) {
        highest = area;
      }
      height[p1] < height[p2] ? p1++ : p2++;
    }

    return highest;
  }
}

void main(List<String> args) {
  String s = "Samuel";
  String john = s[2];
}
